enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -407,7 +407,82 @@ TODO
 
 ### Kubernetes Garbage Collection extended info
 
-TODO
+This feature is relying on Kubernetes garbage collection to delete the
+persistent IP allocations once the Virtual Machine object is deleted.
+
+But, we actually need to assure the persistent IP allocation stays around as
+long as the running Virtual Machine (i.e. the process) is running - thus, we
+also need to ensure the pod where the VM runs was deleted. If it weren't, we
+will end up in a situation where the IP would be returned to the pool earlier
+than it should, which would lead to the same IP in two workloads at the same
+time: one of them shutting down, the other already readily processing traffic.
+
+The garbage collection mechanism will behave differently depending on the
+[DeletionPropagation](https://github.com/kubernetes/apimachinery/blob/3e2600dc79feea6cdc8a9224bc8a6a7fcfee1466/pkg/apis/meta/v1/types.go#L526)
+configured by the user on the `DeleteOptions` provided when deleting the VM.
+
+#### Deleting the VM with the Background `DeletionPropagation`
+
+The following sequence diagram describes how Kubernetes GC works for the
+KubeVirt entities (VM / VMI / pod):
+
+```mermaid
+sequenceDiagram
+  actor User
+  participant persistentIP
+  participant GC
+  participant VM
+  participant VMI
+  participant launcherPod
+
+  User->>VM: Delete VM
+  VM->>VMI: Delete VMI
+  activate VM
+  VMI->>launcherPod: Delete launcherPod
+  activate VMI
+  launcherPod-->>VMI: launcherPod Deleted
+  Deactivate VMI
+  note over VMI: clear VMI finalizer
+  VMI-->>VM: VMI Deleted
+  Deactivate VM
+  note over VM: clear VM finalizer
+  VM-->>User: VM Deleted
+  VM->>GC: VM deleted
+  GC->>persistentIP: Delete persistentIP
+  persistentIP-->>GC: persistentIP Deleted
+```
+
+The VM has a finalizer that prevents it from being deleted while the Virtual
+Machine Instance (VMI, represents a running VM) is around; the VMI - in turn -
+has a finalizer that prevents it from being deleted until the pod where the VM
+runs is deleted. These two ensure the VM object is only removed from the
+datastore after the VM has actually stopped. Thus, in this scenario, we have
+avoided the potential issue described above.
+
+#### Deleting the VM with the Foreground `DeletionPropagation`
+
+When using the foreground `DeletionPropagation`, the dependent objects - e.g.
+the VMI, and the persistentIP allocations - will be deleted **before** the VM.
```

Comment:
I would suggest maybe to rephrase it

There are two trees here:

1.The VM - VMI - Pod, that thanks to the proprietary finalizers will leave VM as the last one that is deleted (same as previous paragraph), regardless the fact we are using foreground which affects children to be deleted immediately, without waiting to become orphaned 
2. The 2nd tree is VM - CRD
this tree is the one that is affected by the deletion policy, worth to emphasize it if possible please

Reply:
Done. Is this better now ? 
@oshoval

Reply:
I think the finalizer explanation should be before diving into propagation types
and then on each propagation tells only what is relevant to the propagation,
having the common part about the finalizers as already known and given

Moreover about this paragraph, i would focus please only about the persistentIP,
because this what matters, and about the pod / VMI we already know (assuming we have a common paragraph as suggested above)

```
When using the foreground `DeletionPropagation`, the dependent object's - e.g.
the VMI, and the persistentIP allocations - deletion will be triggered
**before** the VM is actually deleted.
```

Reply:
Do you agree the diagram represents background deletion ? 

IMO it is easier to describe the flow having the diagram as a crutch. Thus I introduce the finalizers with the diagram. 

You're right that this whole text must be read in sequence, which is not optimal.

Furthermore, the suggested structure is:
- first heading: explain the problem. 
- second heading, background deletion: explain why there's no problem here.
- second heading, foreground deletion: explain why there's a problem here, not with VM/VMI/pod deletion, but with PersistentIP deletion
- ...

Does it make sense in this context ? I've tried to rewrite it with your suggestion and I simply couldn't make the diagram fit.

Reply:
I think that the diagram can come simply at the end, saying that it depicts the supported delete option "background", or simply under background heading as it is now.

Personally i find some redundancy and info which is not the essence in the current headings

The only diff i think is needed is to more the part about finalizers, to some pre heading exposition paragraph,
and then to focus each heading on what only matters there, wdyt?

Reply:
Done. How about now ?

Reply:
Great, thanks
just worth to mention please that background is the default deletion
(and to squash the comment fixing commits at the end :) )

Reply:
Done.